### PR TITLE
ENH: Limited window function support for pandas

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

**Comment by cpcloud** [comment]
This will fail until I add the required data sources to google cloud

File: ibis/sql/postgres/compiler.py [hunk 1]
```diff
@@ -654,11 +654,16 @@ def _floor_divide(t, expr):
...
+    ops.RowNumber: fixed_arity(sa.func.row_number, 0),
     ops.DenseRank: fixed_arity(lambda arg: sa.func.dense_rank(), 1),
     ops.MinRank: fixed_arity(lambda arg: sa.func.rank(), 1),
     ops.PercentRank: fixed_arity(lambda arg: sa.func.percent_rank(), 1),
     ops.NTile: _ntile,
+    ops.FirstValue: fixed_arity(sa.func.first_value, 1),
+    ops.LastValue: fixed_arity(sa.func.first_value, 1),
+    ops.NthValue: fixed_arity(sa.func.nth_value, 2),
+    ops.Lead: fixed_arity(sa.func.lead, 1),
+    ops.Lag: fixed_arity(sa.func.lag, 1),
```
Comment (cpcloud) [inline ibis/sql/postgres/compiler.py]: I'm going to revert because there are no tests. I'll in a follow up PR.

**Comment by cpcloud** [comment]
@wesm can you review this when you get a chance? most important things here are the changes in `ibis/pandas/execution.py` and `ibis/pandas/tests/test_operations.py`.

File: ibis/pandas/execution.py [hunk 2]
```diff
@@ -270,30 +272,51 @@ def execute_cast_string_literal(op, data, type, scope=None):
...
+
+    result = df.assign(**new_columns)
+    result = result.sort_values(computed_sort_keys, ascending=ascending)
+    result = result.drop(new_columns.keys(), axis=1)
+    return result
+
+
 @execute_node.register(ops.Selection, pd.DataFrame)
-def execute_selection_dataframe(op, data, scope=None):
+def execute_selection_dataframe(op, data, scope=None, **kwargs):
```
Comment (cpcloud) [inline ibis/pandas/execution.py]: I need to refactor this function a bit after this PR is merged. It's getting a bit too big for its britches.

**Comment by wesm** [comment]
Wow, big PR! I will take some time to review, am going to try to make a little more progress on the remaining Arrow 0.6.0 issues, and then I can get to this

**Comment by cpcloud** [comment]
One thing to note here is that for the `groupby.transform` style operations (`OVER (PARTITION BY)` in SQL parlance), we execute this in the fastest way possible: by executing each transform separately, rather than executing the operation per group.

Take z-score for example. There are two ways we could implement this.

#### 1. Call `execute` in a `lambda` that executes the operation on each group.

```python
df.groupby('foo').bar.transform(lambda x: (x - x.mean()) / x.std())
```
This will be very slow for high cardinality grouping keys, since pandas can't know that our lambda is composed of a bunch of function calls that are very fast when done individually. Also, the left operand of the subtraction in this case can be taken from the parent `DataFrame`. This avoids any indexing that happens when doing `groupby.apply`.

#### 2. Execute each aggregation in a separate transform.

```python
gb = df.groupby('foo').bar
result = (df.x - gb.transform('mean')) / gb.transform('std')
```
This can be up to 500x (!) faster than the former for certain operations.

This patch always uses the latter when executing any non trivial operations using `t.groupby().mutate()` expressions.

**Comment by cpcloud** [comment]
GCS keeps intermittently failing, rebuilding.

**Comment by cpcloud** [comment]
I'm going to pluck out the `scope` containing `op` instead of `expr` changes here

File: ibis/pandas/execution.py [hunk 3]
```diff
@@ -461,9 +492,19 @@ def execute_reduction_series_groupby_mask(op, data, mask, **kwargs):
...
 
 
+@execute_node.register(ops.Variance, SeriesGroupBy, SeriesGroupBy)
+def execute_reduction_series_groupby_mask_var(op, data, mask, **kwargs):
+    return data.apply(lambda x, mask=mask.obj: x[mask[x.index]].var())
+
+
+@execute_node.register(ops.StandardDev, SeriesGroupBy, SeriesGroupBy)
+def execute_reduction_series_groupby_mask_std(op, data, mask, **kwargs):
+    return data.apply(lambda x, mask=mask.obj: x[mask[x.index]].std())
```
Comment (cpcloud) [inline ibis/pandas/execution.py]: These both should use `context`.

File: ibis/pandas/execution.py [hunk 4]
```diff
@@ -872,3 +913,175 @@ def execute_array_collect(op, data, **kwargs):
...
+
+
+@execute_node.register(ops.CumulativeMin, pd.Series)
+def execute_series_cummin(op, data, **kwargs):
+    return data.cummin()
+
+
+@execute_node.register(ops.CumulativeMax, pd.Series)
+def execute_series_cummax(op, data, **kwargs):
+    return data.cummax()
```
Comment (cpcloud) [inline ibis/pandas/execution.py]: Need to add a test to make sure these do or don't work on `SeriesGroupBy` as well.

**Comment by wesm** [comment]
Reviewing this now. Sorry about the delay

File: ibis/pandas/tests/test_operations.py [hunk 5]
```diff
@@ -1193,3 +1189,247 @@ def test_quantile_array_access(client, t, df):
...
+def test_batting_rolling(batting, batting_df):
+    expr = batting.mutate(
+        more_values=lambda t: t.G.sum().over(
+            ibis.window(order_by=t.yearID, preceding=5)
+        )
+    )
+    result = expr.execute()
+
+    columns = ['G', 'yearID']
+    more_values = batting_df[columns].sort_values('yearID').G.rolling(5).sum()
```
Comment (wesm) [inline ibis/pandas/tests/test_operations.py]: Do you think this will be able to easily handle temporal expressions like "5 days" here?

**Comment by wesm** [comment]
LGTM modulo comment about internal terminology

File: ibis/pandas/context.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
+import abc
+import operator
+
+import six
+
+
+@six.add_metaclass(abc.ABCMeta)
+class Context(object):
```
Comment (cpcloud) [inline ibis/pandas/context.py]: Yeah this isn't named very well.

The idea this encapsulates is really something like `AggregationContext`.

For any particular aggregation such as `sum`, `mean`, etc we need to decide based on the presence or absence of other expressions like `group_by` and `order_by` whether we should call `groupby(...).transform`, `groupby(...).apply`, `groupby(...).expanding().<method>`, `groupby(...).rolling().<method>` or just `groupby(...).<method>` on the data.

Comment (wesm) [inline ibis/pandas/context.py]: Gotcha. The name doesn't matter too much for now as long as it's explained what the thing is.

File: ibis/pandas/tests/test_operations.py (hunk 5 above)
Comment (cpcloud) [inline ibis/pandas/tests/test_operations.py]: I think we can extend `ibis.window` to accept time intervals (probably implemented as an `interval` type) which would be translated to pandas objects and passed into `rolling` which I accepts offsets as of 0.19.0

File: ibis/pandas/context.py (hunk 6 above)
Comment (cpcloud) [inline ibis/pandas/context.py]: I'll add documentation to explain this module and it's classes

Comment (wesm) [inline ibis/pandas/context.py]: :+1:

**Comment by cpcloud** [comment]
Merging on green.

**Pull request closed without merging**